### PR TITLE
Timeout configuration for graphql client, and bugfix

### PR DESCRIPTION
**Pull request by harts75**
Added configuration for setting the timeout request configuration for the http client which was not possible, and consumed the response when the status is not 200. This caused the connections to be not cleaned up, and flood the connectionpool

## Description

1. Added requestConfig for the http client to set the socket, connection and pool timeout
2. Added a line to consume the response, when it is not status 200.

## Related Issue

https://github.com/adobe/commerce-cif-connector/issues/48

## Motivation and Context
When the magento graphql service is malfunctioning and returning http error states, these connections will remain in the pool, until it is full, locking the client. Also added configuration to set timeouts in the service

## How Has This Been Tested?
Used a apache service to return responses with a error code. 

## Types of changes

- [X] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

Is it a feature to prevent a bug or a bugfix. Let's check both.

## Checklist:

- [X] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [X] My change requires a change to the documentation.
-> Update documentation for the configuration of the graphql-client to add timeout values.
- [ ] I have updated the documentation accordingly.
- [X] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes and the overall coverage did not decrease.
Not sure, since the changes are configuration for the http client, which is mocked in the tests.
Can you help out?
- [ ] All unit tests pass on CircleCi.
- [X] I ran all tests locally and they pass.

### CODE REVIEW

**Comment by cjelger** [comment]
Thank you for your PR. I will take some time tomorrow to carefully review that change and also add some tests to make sure that the problem cannot be re-introduced in the future.

**Comment by harts75** [comment]
Hi Christophe,

Thanks for your swift response and action, it is very much appreciated.
I’ll keep an eye out for your changes and tests.

Best regards,
  Harmen.

**Comment by cjelger** [comment]
I finally managed to write some unit tests where I could reproduce the issue. Note that it's quite a nasty bug, because the problem only occurs when the HTTP error response has a body. I will continue with the tests to make sure we fully understand and cover all the issues solved by your PR.  I will probably provide you with feedback tomorrow.

**Comment by harts75** [comment]
Hi Christophe,

I am adding a test to cover the extra lines in the introduced IOException catch, and added
some extra asserts to verify the getentity is called, I am about to push these changes.
I still need to add something to verify the requestconfig is added to the http client.

Best regards,
  Harmen.

File: src/main/java/com/adobe/cq/commerce/graphql/client/impl/GraphqlClientConfiguration.java [hunk 1]
```diff
@@ -23,11 +23,15 @@
...
-    public String DEFAULT_IDENTIFIER = "default";
+    String CQ_GRAPHQL_CLIENT = "cq:graphqlClient";
+    String DEFAULT_IDENTIFIER = "default";
 
-    public static final int MAX_HTTP_CONNECTIONS_DEFAULT = 20;
-    public static final boolean ACCEPT_SELF_SIGNED_CERTIFICATES = false;
+    int MAX_HTTP_CONNECTIONS_DEFAULT = 20;
+    boolean ACCEPT_SELF_SIGNED_CERTIFICATES = false;
+
+    int DEFAULT_CONN_TIMEOUT = 30000;
```
Comment (cjelger) [inline src/main/java/com/adobe/cq/commerce/graphql/client/impl/GraphqlClientConfiguration.java]: I would change this to `5000` (5 seconds), 30 seconds is by far too large.

File: src/main/java/com/adobe/cq/commerce/graphql/client/impl/GraphqlClientConfiguration.java [hunk 2]
```diff
@@ -61,4 +65,22 @@
         description = "The maximum number of concurrent HTTP connections the connector can make",
         type = AttributeType.INTEGER)
     int maxHttpConnections() default MAX_HTTP_CONNECTIONS_DEFAULT;
+
+    @AttributeDefinition(
+        name = "Http connection timeout",
+        description = "The timeout in milliseconds for the http request",
```
Comment (cjelger) [inline src/main/java/com/adobe/cq/commerce/graphql/client/impl/GraphqlClientConfiguration.java]: I'd reuse the text from the http client documentation: `The timeout in milliseconds until a connection is established. A timeout value of zero is interpreted as an infinite timeout.`

File: src/main/java/com/adobe/cq/commerce/graphql/client/impl/GraphqlClientConfiguration.java [hunk 3]
```diff
@@ -61,4 +65,22 @@
...
+        type = AttributeType.INTEGER)
+    int conn_timeout() default DEFAULT_CONN_TIMEOUT;
+
+    @AttributeDefinition(
+        name = "Http socket timeout",
+        description = "The timeout in milliseconds for the socket",
```
Comment (cjelger) [inline src/main/java/com/adobe/cq/commerce/graphql/client/impl/GraphqlClientConfiguration.java]: Same: `The socket timeout in milliseconds, which is the timeout for waiting for data or, put differently, a maximum period inactivity between two consecutive data packets. A timeout value of zero is interpreted as an infinite timeout.`

File: src/main/java/com/adobe/cq/commerce/graphql/client/impl/GraphqlClientConfiguration.java [hunk 4]
```diff
@@ -61,4 +65,22 @@
...
+        type = AttributeType.INTEGER)
+    int so_timeout() default DEFAULT_SO_TIMEOUT;
+
+    @AttributeDefinition(
+        name = "Request pool timeout",
+        description = "The timeout in milliseconds for the connectionpool to return a connection.",
```
Comment (cjelger) [inline src/main/java/com/adobe/cq/commerce/graphql/client/impl/GraphqlClientConfiguration.java]: Same: `The timeout in milliseconds used when requesting a connection from the connection manager. A timeout value of zero is interpreted as an infinite timeout.`

File: src/main/java/com/adobe/cq/commerce/graphql/client/impl/GraphqlClientImpl.java [hunk 5]
```diff
@@ -14,20 +14,17 @@
...
 
 import javax.net.ssl.SSLContext;
 
-import org.apache.http.Header;
-import org.apache.http.HttpEntity;
-import org.apache.http.HttpHeaders;
-import org.apache.http.HttpResponse;
-import org.apache.http.HttpStatus;
-import org.apache.http.StatusLine;
+import org.apache.http.*;
```
Comment (cjelger) [inline src/main/java/com/adobe/cq/commerce/graphql/client/impl/GraphqlClientImpl.java]: Please do not reformat import statements with the `*` wildcard, we explicitly do not want to use wildcard imports.

File: src/main/java/com/adobe/cq/commerce/graphql/client/impl/GraphqlClientImpl.java [hunk 6]
```diff
@@ -71,6 +64,9 @@
     private boolean acceptSelfSignedCertificates;
     private int maxHttpConnections;
     private HttpMethod httpMethod;
+    private int conn_timeout;
```
Comment (cjelger) [inline src/main/java/com/adobe/cq/commerce/graphql/client/impl/GraphqlClientImpl.java]: Please use correct camelcase names for Java variables (I'd also use the same names in the OSGi config, but you can also keep the `_` separator in the OSGi config if you want to, because this is a special allowed syntax in that particular case).

File: src/main/java/com/adobe/cq/commerce/graphql/client/impl/GraphqlClientImpl.java [hunk 7]
```diff
@@ -127,6 +126,11 @@ public String getIdentifier() {
 
             return response;
         } else {
+            try {
+                EntityUtils.consume(httpResponse.getEntity());
```
Comment (cjelger) [inline src/main/java/com/adobe/cq/commerce/graphql/client/impl/GraphqlClientImpl.java]: I'd rather use `EntityUtils.consumeQuietly()` to remove the try-catch, we don't really care here if an exception occurs because we anyway don't use the consumed entity.

File: src/test/java/com/adobe/cq/commerce/graphql/client/impl/GraphqlClientImplTest.java [hunk 8]
```diff
@@ -49,8 +49,9 @@
 import com.google.gson.JsonElement;
 import com.google.gson.JsonParseException;
 
-import static org.junit.Assert.assertEquals;
```
Comment (cjelger) [inline src/test/java/com/adobe/cq/commerce/graphql/client/impl/GraphqlClientImplTest.java]: While it's nice that you added these tests, I'd rather completely revert these changes. I actually created new unit tests to test the concurrency of the client: these tests will fail if the HTTP entities are not consumed, without having to rely on checking implementation details.

File: src/main/java/com/adobe/cq/commerce/graphql/client/impl/GraphqlClientConfiguration.java (hunk 1 above)
Comment (harts75) [reply src/main/java/com/adobe/cq/commerce/graphql/client/impl/GraphqlClientConfiguration.java]: Agree and done

File: src/main/java/com/adobe/cq/commerce/graphql/client/impl/GraphqlClientConfiguration.java (hunk 2 above)
Comment (harts75) [reply src/main/java/com/adobe/cq/commerce/graphql/client/impl/GraphqlClientConfiguration.java]: Done

File: src/main/java/com/adobe/cq/commerce/graphql/client/impl/GraphqlClientConfiguration.java (hunk 3 above)
Comment (harts75) [reply src/main/java/com/adobe/cq/commerce/graphql/client/impl/GraphqlClientConfiguration.java]: Done

File: src/main/java/com/adobe/cq/commerce/graphql/client/impl/GraphqlClientImpl.java (hunk 5 above)
Comment (harts75) [reply src/main/java/com/adobe/cq/commerce/graphql/client/impl/GraphqlClientImpl.java]: Done

File: src/main/java/com/adobe/cq/commerce/graphql/client/impl/GraphqlClientImpl.java (hunk 7 above)
Comment (harts75) [reply src/main/java/com/adobe/cq/commerce/graphql/client/impl/GraphqlClientImpl.java]: Done

File: src/test/java/com/adobe/cq/commerce/graphql/client/impl/GraphqlClientImplTest.java (hunk 8 above)
Comment (harts75) [reply src/test/java/com/adobe/cq/commerce/graphql/client/impl/GraphqlClientImplTest.java]: Much better indeed. Other test changes have been reverted.

File: src/main/java/com/adobe/cq/commerce/graphql/client/impl/GraphqlClientImpl.java (hunk 6 above)
Comment (harts75) [reply src/main/java/com/adobe/cq/commerce/graphql/client/impl/GraphqlClientImpl.java]: done

File: src/main/java/com/adobe/cq/commerce/graphql/client/impl/GraphqlClientConfiguration.java (hunk 4 above)
Comment (harts75) [reply src/main/java/com/adobe/cq/commerce/graphql/client/impl/GraphqlClientConfiguration.java]: done

File: src/test/java/com/adobe/cq/commerce/graphql/client/impl/GraphqlClientImplTest.java (hunk 8 above)
Comment (cjelger) [inline src/test/java/com/adobe/cq/commerce/graphql/client/impl/GraphqlClientImplTest.java]: Please revert all changes to this file.

File: src/main/java/com/adobe/cq/commerce/graphql/client/impl/GraphqlClientImpl.java [hunk 9]
```diff
@@ -21,13 +21,9 @@
 
 import javax.net.ssl.SSLContext;
 
-import org.apache.http.Header;
-import org.apache.http.HttpEntity;
-import org.apache.http.HttpHeaders;
-import org.apache.http.HttpResponse;
-import org.apache.http.HttpStatus;
-import org.apache.http.StatusLine;
+import org.apache.http.*;
```
Comment (cjelger) [inline src/main/java/com/adobe/cq/commerce/graphql/client/impl/GraphqlClientImpl.java]: Please revert the use of wildcard import.

File: src/main/java/com/adobe/cq/commerce/graphql/client/impl/GraphqlClientConfiguration.java [hunk 10]
```diff
@@ -61,4 +65,22 @@
         description = "The maximum number of concurrent HTTP connections the connector can make",
         type = AttributeType.INTEGER)
     int maxHttpConnections() default MAX_HTTP_CONNECTIONS_DEFAULT;
+
+    @AttributeDefinition(
+        name = "Http connection timeout",
+        description = "The timeout in milliseconds until a connection is established. A timeout value of zero is interpreted as an infinite timeout.",
+        type = AttributeType.INTEGER)
+    int connTimeout() default DEFAULT_CONN_TIMEOUT;
```
Comment (cjelger) [inline src/main/java/com/adobe/cq/commerce/graphql/client/impl/GraphqlClientConfiguration.java]: I'm being picky here, but better use full names, they will be easier to read for anyone new wondering in a few months what "conn" might mean: `connectionTimeout`. Please also apply the same changes for the variables in the `GraphqlClientImpl` class, thx.

File: src/main/java/com/adobe/cq/commerce/graphql/client/impl/GraphqlClientConfiguration.java [hunk 11]
```diff
@@ -61,4 +65,22 @@
...
+
+    @AttributeDefinition(
+        name = "Http socket timeout",
+        description = "The socket timeout in milliseconds, which is the timeout for waiting for data or, put differently, a maximum period inactivity between two consecutive data packets. A timeout value of zero is interpreted as an infinite timeout.",
+        type = AttributeType.INTEGER)
+    int soTimeout() default DEFAULT_SO_TIMEOUT;
```
Comment (cjelger) [inline src/main/java/com/adobe/cq/commerce/graphql/client/impl/GraphqlClientConfiguration.java]: `socketTimeout`

File: src/main/java/com/adobe/cq/commerce/graphql/client/impl/GraphqlClientConfiguration.java [hunk 12]
```diff
@@ -61,4 +65,22 @@
...
+
+    @AttributeDefinition(
+        name = "Request pool timeout",
+        description = "The timeout in milliseconds used when requesting a connection from the connection manager. A timeout value of zero is interpreted as an infinite timeout.",
+        type = AttributeType.INTEGER)
+    int reqpoolTimeout() default DEFAULT_REQPOOL_TIMEOUT;
```
Comment (cjelger) [inline src/main/java/com/adobe/cq/commerce/graphql/client/impl/GraphqlClientConfiguration.java]: `requestPoolTimeout`

File: src/main/java/com/adobe/cq/commerce/graphql/client/impl/GraphqlClientImpl.java (hunk 9 above)
Comment (cjelger) [reply src/main/java/com/adobe/cq/commerce/graphql/client/impl/GraphqlClientImpl.java]: Please just fix the wildcard and I think it will all be fine. Thx.

Comment (harts75) [reply src/main/java/com/adobe/cq/commerce/graphql/client/impl/GraphqlClientImpl.java]: Totally missed that one. Thanks for all help and input, very much appreciated

**Comment by harts75** [comment]
Thanks again for all your input and helping to get it merged.
I looked at the way you are testing this, and this looks nice. Will be incorporating it in our own test cases too if I find the chance.
Can you maybe give an indication of the release cycle for the graphql client, and maybe also for the commerce-cif-connector?

**Comment by cjelger** [comment]
I will test everything again with the GraphQL connector, and will probably release a new version of the client tomorrow, and then update the dependency in the connector. I don't yet for the release of the connector itself, but at least you'll be able to use the SNAPSHOT version until we also release it.

**Comment by harts75** [comment]
Ok, thanks for the info, that’s quick.
I’ll keep an eye out.
